packages/pwa/app/commerce-api/contexts.js

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import React, {createContext, useContext, useReducer} from 'react'
```

Comment:
With respect to using the entire action/reducer pattern, I thought we had a conversation that concluded such that we thought it was a good idea, but maybe to hold off until we refactor the commerce api into the sdk as the pattern doesn't match how our other context/providers are working?

At a minimum I would like to see a ticket to adjust the other context/providers to use reducers. That would serve as pre-work to the refactor for commerce api.

Reply:
(continued from our gmeet conversation)

The existing contexts that we have are pretty simple, just a value and a setValue function to update the entire context. For example for categories, we fetch the categories and put them in the context, that's it. But for customer product lists, we had to do more than that, partly due to the limitation of the API, we need to implement logic to plug product details in the the product lists, and having that in the actions (hooks) are hard to maintain. We didn't need to implement actions/reducers until now.

You mentioned in the call that we want to have "thick" providers and "thin" hooks, that a user can choose to use either one of them. I didn't have that mental model and probably need some time to think about it. In my head, the structure was like "Provider -> hook -> UI component" kind of thing for me, there is no direct relationship between the provider and the UI components. (I'm not saying this is the direction we should go)

In this PR, the current customer product list provider is implemented in a way that should not be used without the hook... The core of the business logic is in the hook not the provider. And this is also how the existing hooks like `useBasket` and `useCustomer` work too.

I'm happy to create tickets for us to think about the structures and you are definitely right this is good exercise for us to prepare to move the commerce api hooks to the SDK.